## ЭТАП 1. Получение данных, визуальный анализ структуры таблицы

In [19]:
import pandas as pd #Обращение к библиотеке pandas

df = pd.read_csv('GENERIC_DICTIONARIES_20151120000000.csv',sep = "|") #Чтение файла
df.head() #Отображение первых 5-ти строк таблицы

H          CREATE_DATE MODIFIED_DATE REMOVED_DATE DICTIONARY_CODE CODE  \
0  I  2015-11-20 09:32:56           NaN          NaN      ACC_STATUS    A   
1  I  2015-11-20 09:32:56           NaN          NaN      ACC_STATUS    S   
2  I  2015-11-20 09:32:56           NaN          NaN      ACC_STATUS    I   
3  I  2015-11-20 09:32:56           NaN          NaN      ACC_STATUS    C   
4  I  2015-11-20 09:32:56           NaN          NaN      ACC_STATUS    D   

            NAME  PARENT_DICT_CODE FUTURE_USE_1  FUTURE_USE_2  FUTURE_USE_3  \
0        Активен               NaN           ru           NaN           NaN   
1  Приостановлен               NaN           ru           NaN           NaN   
2      Неактивен               NaN           ru           NaN           NaN   
3         Закрыт               NaN           ru           NaN           NaN   
4           Умер               NaN           ru           NaN           NaN   

   FUTURE_USE_4  FUTURE_USE_5  FUTURE_USE_6  FUTURE_USE_7  FUTURE_USE_8  \
0           NaN           NaN           NaN           NaN           NaN   
1           NaN           NaN           NaN           NaN           NaN   
2           NaN           NaN           NaN           NaN           NaN   
3           NaN           NaN           NaN           NaN           NaN   
4           NaN           NaN           NaN           NaN           NaN   

   FUTURE_USE_9 FUTURE_USE_10,,,,  
0           NaN              ,,,,  
1           NaN              ,,,,  
2           NaN              ,,,,  
3           NaN              ,,,,  
4           NaN              ,,,,

In [20]:
df.columns #Определение переченя названий столбцов таблицы df

Index(['H', 'CREATE_DATE', 'MODIFIED_DATE', 'REMOVED_DATE', 'DICTIONARY_CODE',
       'CODE', 'NAME', 'PARENT_DICT_CODE', 'FUTURE_USE_1', 'FUTURE_USE_2',
       'FUTURE_USE_3', 'FUTURE_USE_4', 'FUTURE_USE_5', 'FUTURE_USE_6',
       'FUTURE_USE_7', 'FUTURE_USE_8', 'FUTURE_USE_9', 'FUTURE_USE_10,,,,'],
      dtype='object')

In [21]:
df = df.loc[:,'H':'FUTURE_USE_1'] #Срез первых 9 столбцов
df.head() #Срез первых 5 строк таблицы


H          CREATE_DATE MODIFIED_DATE REMOVED_DATE DICTIONARY_CODE CODE  \
0  I  2015-11-20 09:32:56           NaN          NaN      ACC_STATUS    A   
1  I  2015-11-20 09:32:56           NaN          NaN      ACC_STATUS    S   
2  I  2015-11-20 09:32:56           NaN          NaN      ACC_STATUS    I   
3  I  2015-11-20 09:32:56           NaN          NaN      ACC_STATUS    C   
4  I  2015-11-20 09:32:56           NaN          NaN      ACC_STATUS    D   

            NAME  PARENT_DICT_CODE FUTURE_USE_1  
0        Активен               NaN           ru  
1  Приостановлен               NaN           ru  
2      Неактивен               NaN           ru  
3         Закрыт               NaN           ru  
4           Умер               NaN           ru

## ЭТАП 2. Подготовка таблицы дл анализа - удаление/замена пропусков, удаление дублей, удаление строк по условиям, переименование столбцов

**Проверка на таблицы на качество** 

In [22]:
df.info() #Получение общей информации о данных в таблице df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9799 entries, 0 to 9798
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   H                 9799 non-null   object 
 1   CREATE_DATE       9799 non-null   object 
 2   MODIFIED_DATE     78 non-null     object 
 3   REMOVED_DATE      128 non-null    object 
 4   DICTIONARY_CODE   9798 non-null   object 
 5   CODE              9798 non-null   object 
 6   NAME              9728 non-null   object 
 7   PARENT_DICT_CODE  0 non-null      float64
 8   FUTURE_USE_1      9798 non-null   object 
dtypes: float64(1), object(8)
memory usage: 689.1+ KB


ВЫВОДЫ проверки таблицы на качество:
1. Колонка "PARENT_DICT_CODE" содержит только NaN значения
2. Колонка "REMOVED_DATE" содержит 128 заполненных значений
2. Колонка "MODIFIED_DATE" содержит 78 заполненных значений
3. Колонка "NAME" содержит 9728 заполненных значений
4. Колонка "FUTURE_USE_1","DICTIONARY_CODE","CODE" содержит одно значение «Nan»

Редактирование таблицы, ЗАДАЧИ:
1. Переименовать колонки: изменить верхний регистр на нижний
4. Заменить «Nan» значения в "removed_date","modified_date" и "name" на '-'
3. Удалить строку, где Колонка "future_use_1"= "Nan"
4. Проверить строки на дубликаты 
5. Удалить дублирующие строки при необходимости

In [23]:
# ЗАДАЧА 1.Переименование колонок: изменение верхнего регистра на нижний
df.set_axis(['h', 'create_date', 'modified_date', 'removed_date', 'dictionary_code',
       'code', 'name', 'parent_dict_code', 'future_use_1'],axis = 'columns',inplace = True)
df.columns

Index(['h', 'create_date', 'modified_date', 'removed_date', 'dictionary_code',
       'code', 'name', 'parent_dict_code', 'future_use_1'],
      dtype='object')

In [24]:
df['removed_date'].isnull().sum() #Сумма значений "Nan" в колонке 'removed_date'

9671

In [25]:
# ЗАДАЧА 2.Замена значений Nan в "removed_date","modified_date" и "name" на '-'
df['removed_date'] = df['removed_date'].fillna('-')
df['modified_date'] = df['modified_date'].fillna('-')
df['name'] = df['name'].fillna('-')
df['parent_dict_code'] = df['parent_dict_code'].fillna('-')

df.head(10) #Проверка

h          create_date modified_date         removed_date dictionary_code  \
0  I  2015-11-20 09:32:56             -                    -      ACC_STATUS   
1  I  2015-11-20 09:32:56             -                    -      ACC_STATUS   
2  I  2015-11-20 09:32:56             -                    -      ACC_STATUS   
3  I  2015-11-20 09:32:56             -                    -      ACC_STATUS   
4  I  2015-11-20 09:32:56             -                    -      ACC_STATUS   
5  I  2015-11-20 09:32:56             -                    -      ACC_STATUS   
6  I  2015-11-20 09:32:56             -                    -      ACC_STATUS   
7  I  2015-11-20 09:32:56             -                    -       CRH_TITLE   
8  I  2015-11-20 09:32:56             -                    -       CRH_TITLE   
9  I  2015-11-20 09:32:56             -  2015-11-23 10:41:50       CRH_TITLE   

  code           name parent_dict_code future_use_1  
0    A        Активен                -           ru  
1    S  Приостановлен                -           ru  
2    I      Неактивен                -           ru  
3    C         Закрыт                -           ru  
4    D           Умер                -           ru  
5    M      Объединен                -           ru  
6    O         Открыт                -           ru  
7   MR            Г-н                -           ru  
8   MS           Г-жа                -           ru  
9  MRS            Mrs                -           ru

In [26]:
# ЗАДАЧА 3.Удаление строки, где колонка "future_use_1"= 'Nan'
df.dropna(subset = ['future_use_1'], inplace = True) #Удаление строки, где в столбце 'future_use_1' содержитсся значение Nan
df.info() #Проверка удаленной строки. Всего было 9899 значений, стало 9798 - все верно.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9798 entries, 0 to 9797
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   h                 9798 non-null   object
 1   create_date       9798 non-null   object
 2   modified_date     9798 non-null   object
 3   removed_date      9798 non-null   object
 4   dictionary_code   9798 non-null   object
 5   code              9798 non-null   object
 6   name              9798 non-null   object
 7   parent_dict_code  9798 non-null   object
 8   future_use_1      9798 non-null   object
dtypes: object(9)
memory usage: 765.5+ KB


In [27]:
# ЗАДАЧА 4. Проверка строк на дубликаты
df.duplicated().sum()

4899

In [28]:
# ЗАДАЧА 5. Удаление дубликатов
df = df.drop_duplicates().reset_index(drop = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4899 entries, 0 to 4898
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   h                 4899 non-null   object
 1   create_date       4899 non-null   object
 2   modified_date     4899 non-null   object
 3   removed_date      4899 non-null   object
 4   dictionary_code   4899 non-null   object
 5   code              4899 non-null   object
 6   name              4899 non-null   object
 7   parent_dict_code  4899 non-null   object
 8   future_use_1      4899 non-null   object
dtypes: object(9)
memory usage: 344.6+ KB


**Все строки заполнены по первым девяти колонкам. Дубликатов нет**

## ЭТАП 3. Анализ данных

In [29]:
df.head(100)

h          create_date modified_date removed_date dictionary_code code  \
0   I  2015-11-20 09:32:56             -            -      ACC_STATUS    A   
1   I  2015-11-20 09:32:56             -            -      ACC_STATUS    S   
2   I  2015-11-20 09:32:56             -            -      ACC_STATUS    I   
3   I  2015-11-20 09:32:56             -            -      ACC_STATUS    C   
4   I  2015-11-20 09:32:56             -            -      ACC_STATUS    D   
.. ..                  ...           ...          ...             ...  ...   
95  I  2015-11-20 09:32:56             -            -        COM_TYPE    1   
96  I  2015-11-20 09:32:56             -            -        COM_TYPE    4   
97  I  2015-11-20 09:32:56             -            -        COM_TYPE    5   
98  I  2015-11-20 09:32:56             -            -        COM_TYPE   FP   
99  I  2015-11-20 09:32:56             -            -        COM_TYPE   AU   

                       name parent_dict_code future_use_1  
0                   Активен                -           ru  
1             Приостановлен                -           ru  
2                 Неактивен                -           ru  
3                    Закрыт                -           ru  
4                      Умер                -           ru  
..                      ...              ...          ...  
95                   Coupon                -           ru  
96  Facebook new elite tier                -           ru  
97           Facebook award                -           ru  
98       Forgotten Password                -           ru  
99    Member profile update                -           ru  

[100 rows x 9 columns]

**Вывод уникальных типов статусов в алфавитном порядке**

In [30]:
sort_status_df= df['dictionary_code'].sort_values(ascending = True) #Сортировка
sort_status_df = sort_status_df.unique() #Поиск уникальных значений
sort_status_df

array(['ACC_STATUS', 'ACT_STATUS', 'ACT_TYPE', 'APP_VARIANT',
       'AWR_STATUS', 'AWR_TYPE', 'BRL_STATUS', 'BRL_TYPE', 'BTC_STATUS',
       'CASE_TYPE', 'COM_CHANNEL', 'COM_STATUS', 'COM_TYPE',
       'CON_CHANNEL', 'CON_STATUS', 'COUNTRIES', 'CRH_GENDER',
       'CRH_TITLE', 'CUS_SUBTYPE', 'CUS_TYPE', 'ENROLL_CHANNEL',
       'MAQ_STATUS', 'MBR_STATUS_CHANGE_SOURCE_CODE', 'NNT_STATUS',
       'NON_STATUS', 'PREF_LANGUAGES', 'PRT_INDUSTRY', 'PRT_STATUS',
       'PRT_TYPE', 'QSN_TYPE', 'QST_CUS_STATUS', 'RDI_STATUS', 'RR_CODE',
       'SEG_STATUS', 'SRV_STATUS', 'STATES', 'TRN_CHANNEL', 'TRN_STATUS',
       'USR_STATUS', 'USR_TYPE', 'VOU_STATUS', 'VOU_TYPE'], dtype=object)

**Подсчет значений 'Name' в каждом типе статусов**

In [31]:
df.groupby('dictionary_code')['name'].count() #Группировка таблицы по 'dictionary_code', подсчет значений по стобцу 'Name' для каждого 'dictionary_code'

dictionary_code
ACC_STATUS                          7
ACT_STATUS                          3
ACT_TYPE                           39
APP_VARIANT                         3
AWR_STATUS                          3
AWR_TYPE                            8
BRL_STATUS                          3
BRL_TYPE                           19
BTC_STATUS                          6
CASE_TYPE                           6
COM_CHANNEL                         3
COM_STATUS                          6
COM_TYPE                           67
CON_CHANNEL                         4
CON_STATUS                          2
COUNTRIES                         249
CRH_GENDER                          2
CRH_TITLE                          20
CUS_SUBTYPE                         4
CUS_TYPE                            2
ENROLL_CHANNEL                      5
MAQ_STATUS                          8
MBR_STATUS_CHANGE_SOURCE_CODE       3
NNT_STATUS                          6
NON_STATUS                          4
PREF_LANGUAGES                    

In [32]:
df.groupby('dictionary_code')['name'].count().mean() #Поиск среднего арифмитеческого количества значений в статусах

116.64285714285714

In [33]:
df.groupby('dictionary_code')['name'].count().max() #Поиск максимального количества значений в статусах

4331

**Вывод таблицы, в которой dictionary_code = STATES**

In [34]:
df_STATES = df.loc[df['dictionary_code'] == 'STATES']
df_STATES.head(10) #Отображение первых 10-ти значений

h          create_date modified_date removed_date dictionary_code  \
523  I  2015-11-20 09:32:56             -            -          STATES   
524  I  2015-11-20 09:32:56             -            -          STATES   
525  I  2015-11-20 09:32:56             -            -          STATES   
526  I  2015-11-20 09:32:56             -            -          STATES   
527  I  2015-11-20 09:32:56             -            -          STATES   
528  I  2015-11-20 09:32:56             -            -          STATES   
529  I  2015-11-20 09:32:56             -            -          STATES   
530  I  2015-11-20 09:32:56             -            -          STATES   
531  I  2015-11-20 09:32:56             -            -          STATES   
532  I  2015-11-20 09:32:56             -            -          STATES   

       code                 name parent_dict_code future_use_1  
523   AD-07     Andorra la Vella                -           ru  
524   AD-02              Canillo                -           ru  
525   AD-03               Encamp                -           ru  
526   AD-08   Escaldes-Engordany                -           ru  
527   AD-04           La Massana                -           ru  
528   AD-05               Ordino                -           ru  
529   AD-06  Sant Julia de Loria                -           ru  
530  AO-BGO                Bengo                -           ru  
531  AO-BGU             Benguela                -           ru  
532  AO-BIE                  Bié                -           ru

## ЭТАП 4. Результаты

In [35]:
df_STATES.to_csv('DF_STATES.csv', sep = ";") #Запись таблицы, в которой dictionary_code = STATES, в новый документ